netadapter/netadapter.go

```diff
@@ -43,31 +61,118 @@ func (na *NetAdapter) Start() error {
 
 // Stop safely closes the NetAdapter
 func (na *NetAdapter) Stop() error {
+	if atomic.AddInt32(&na.stop, 1) != 1 {
+		log.Warnf("Net adapter stopped more than once")
+		return nil
+	}
 	return na.server.Stop()
```

Comment:
I don't think we want to `Stop` na.server twice either, bad things might happen.
So we need to either add the atomic stuff to server and anything under it as well, or insert the na.server.Stop() into the if. I prefer the later